pkg/server/event_handler.go

```diff
@@ -234,33 +241,88 @@ func (h *EventHandler[T, P]) HTTPHandler() http.Handler {
 // [GCS notification]: https://cloud.google.com/storage/docs/pubsub-notifications#format
 func (h *EventHandler[T, P]) Handle(ctx context.Context, m pubsub.Message) error {
 	logger := logging.FromContext(ctx)
+
+	eventBytes, processErr := h.generatePmapEventBytes(ctx, m)
+
+	attr := map[string]string{}
+
+	// Use successMessenger to send to downstream if there are no
+	// errors for generating a PmapEvent.
+	// if processErr == nil {
+	// 	if err := h.successMessenger.Send(ctx, eventBytes, attr); err != nil {
+	// 		return fmt.Errorf("failed to send succuss event downstream: %w", err)
+	// 	}
+	// 	return nil
+	// }
+
+	// fmt.Println("processErr: " + processErr.Error())
+
+	// If error is a retryable error, we return err
+	// so we can start retry.
+	if processErr != nil {
+		if !errors.Is(processErr, pmaperrors.ErrNonRetryable) {
+			return processErr
+		}
+		attr[nonRetryableError] = processErr.Error()
+		logger.Errorw(processErr.Error(), "bucketId", m.Attributes["bucketId"], "objectId", m.Attributes["objectId"])
+		if err := h.failureMessenger.Send(ctx, eventBytes, attr); err != nil {
+			fmt.Println("Failure failed to send downstream")
+			return fmt.Errorf("failed to send failure event downstream: %w", err)
+		}
+		return nil
+	}
+	if err := h.successMessenger.Send(ctx, eventBytes, attr); err != nil {
+		return fmt.Errorf("failed to send succuss event downstream: %w", err)
+	}
+	return nil
+	// if errors.Is(processErr, pmaperrors.ErrRetryable) {
+	// 	fmt.Println("THIS IS RETRYABLE ERROR")
+	// 	return fmt.Errorf("EventHandler encountered retryable error: %w", processErr)
+	// }
+
+	// If the error is not retryable, we use failerMessenger
+	// to send the message to corresponding topic.
+	// attr[nonRetryableError] = processErr.Error()
+	// logger.Errorw(processErr.Error(), "bucketId", m.Attributes["bucketId"], "objectId", m.Attributes["objectId"])
+	// if err := h.failureMessenger.Send(ctx, eventBytes, attr); err != nil {
+	// 	fmt.Println("Failure failed to send downstream")
+	// 	return fmt.Errorf("failed to send failure event downstream: %w", err)
+	// }
+	// This is supposed to return nil,
+	// currently, all errors are not of type ErrRetryable
+	// but we handle them as retryable errors.
+	// We will change this to nil after we sort out
+	// how to differenate these errors.
+	// return nil
+}
+
+func (h *EventHandler[T, P]) generatePmapEventBytes(ctx context.Context, m pubsub.Message) ([]byte, error) {
```

Comment:
I thought you need to return the error either it is nontreyable or retryable error?